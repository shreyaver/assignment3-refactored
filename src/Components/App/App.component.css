.app {
  display: flex;
  flex-direction: column;
  min-height: 100%;
}
.loadingStatus {
  flex: 1;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  font-size: 2rem;
  color: var(--white);
  font-weight: bold;
  font-style: italic;
  text-align: center;
}