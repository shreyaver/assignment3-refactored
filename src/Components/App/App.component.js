import React, { Component } from 'react';
import Axios from 'axios';
import Header from '../Header/Header.component';
import Books from '../Books/Books.component';
import { LIKE_BOOK_URL } from '../../config';
import getBooksAndLiked from '../../Helpers/getBooksAndLiked/getBooksAndLiked';
import './App.component.css';

class App extends Component {
  state = {};

  async componentDidMount() {
    this.setState(await getBooksAndLiked());
  }

  handleHeaderClick = async () => {
    this.setState({ books: undefined });
    await this.componentDidMount();
  }
 
  handleLikeClick = book => async () => {
    const serverResponse = await Axios.post(`${LIKE_BOOK_URL}/${book.id}`, { liked: !book.liked });
    if ((serverResponse.data.indexOf('Book liked!') !== -1) || (serverResponse.data.indexOf('Book disliked!') !== -1)) {
      this.setState(state => ({
        books: {
          ...state.books,
          [book.Author]: state.books[book.Author].map((bookObj) => {
            if (bookObj.id !== book.id) {
              return { ...bookObj };
            }
            return { ...bookObj, liked: !bookObj.liked };
          }),
        },
      }));
    }
  };

  render() {
    const { books } = this.state;
    if (books === undefined) {
      return (
        <div className="app">
          <Header handleHeaderClick={this.handleHeaderClick} />
          <div className="loadingStatus">
            LOADING ...
          </div>
        </div>
      );
    }
    return (
      <div className="app">
        <Header handleHeaderClick={this.handleHeaderClick} />
        <Books books={books} handleLikeClick={this.handleLikeClick} />
      </div>
    );
  }
}

export default App;
