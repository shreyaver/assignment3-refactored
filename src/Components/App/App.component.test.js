import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Axios from 'axios';
import App from './App.component';
import mockBooksWithLiked from '../../mockData/booksWithLiked.json';

jest.mock('../../Helpers/getBooksAndLiked/getBooksAndLiked', () => jest.fn().mockImplementation(() => Promise.resolve(mockBooksWithLiked)));

// testing header refresh?
// testing snapshot after loading?
let mockLikeBook;

beforeAll(() => {
  mockLikeBook = jest.spyOn(Axios, 'post');
  mockLikeBook.mockImplementation((url, likedPayload) => Promise.resolve({ data: likedPayload.liked === true ? 'Book liked!' : 'Book disliked!' }));
});
afterAll(() => {
  jest.restoreAllMocks();
});
describe('App', () => {
  it('renders for loading page', () => {
    const loadingPage = renderer.create(<App />).toJSON();
    expect(loadingPage).toMatchSnapshot();
  });

  // it('renders after data is fetched', async () => {
  //   const app = renderer.create(<App />);
  //   await app.toTree().instance.componentDidMount();
  //   expect(app.toJSON()).toMatchSnapshot();
  // });

  it('sets initial state to empty object', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.instance().state).toEqual({});
  });

  it('sets state to fetched data after mounting', async () => {
    const wrapper = shallow(<App />);
    await wrapper.instance().componentDidMount();
    expect(wrapper.instance().state).toEqual(mockBooksWithLiked);
  });
});

describe('handleLikeClick()', () => {
  it('changes liked state for passed book', async () => {
    const wrapper = shallow(<App />);
    await wrapper.instance().componentDidMount();
    const originalBook = Object.values(wrapper.instance().state.books)[0][0];
    await wrapper.instance().handleLikeClick(originalBook)();
    expect(Object.values(wrapper.instance().state.books)[0][0]).toEqual({ ...originalBook, liked: !originalBook.liked });
  });
});
