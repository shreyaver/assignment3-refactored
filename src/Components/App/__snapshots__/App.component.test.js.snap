// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders for loading page 1`] = `
<div
  className="app"
>
  <header
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="headerLine"
    >
      <hr />
    </div>
    <div
      className="headerTitle"
    >
      <div
        className="circle"
      >
        BS
      </div>
      <div
        className="headerCaption"
      >
        The Book Shelf
      </div>
    </div>
    <div
      className="headerLine"
    >
      <hr />
    </div>
  </header>
  <div
    className="loadingStatus"
  >
    LOADING ...
  </div>
</div>
`;
